### PR TITLE
Add Quarantine doc

### PR DESCRIPTION
**Pull request by fgimenez**
Signed-off-by: Federico Gimenez <fgimenez@redhat.com>

**What this PR does / why we need it**: It adds a description of the test quarantine procedure.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: docs/quarantine.md
Comment (dankenigsberg) [inline docs/quarantine.md:7]:
```suggestion
* The results of the whole suite are less valuable, we can no longer trust
```

File: docs/quarantine.md
Comment (dankenigsberg) [inline docs/quarantine.md:9]: How is this different from the former point?

File: docs/quarantine.md
Comment (dankenigsberg) [inline docs/quarantine.md:11]:
```suggestion
* The development process is affected negatively. Developers need to wait
```

File: docs/quarantine.md
Comment (dankenigsberg) [inline docs/quarantine.md:29]:
```suggestion
In the following definitions we will consider test failures only when reported on
```
(Trying to his your meaning here)

File: docs/quarantine.md
Comment (dankenigsberg) [inline docs/quarantine.md:36]:
```suggestion
* It has failed 2 times or more in the last two weeks.
```
I'm not sure about this English suggestions.

File: docs/quarantine.md
Comment (dankenigsberg) [inline docs/quarantine.md:37]: You already wrote this in the preface. We need to say this only once.

File: docs/quarantine.md
Comment (dankenigsberg) [inline docs/quarantine.md:40]:
```suggestion
A PR will be proposed on Mondays every two weeks with a batch of the tests that
```

File: docs/quarantine.md
Comment (dankenigsberg) [inline docs/quarantine.md:44]:
```suggestion
Each quarantined test must have a team owner. The PR will add the text
```

File: docs/quarantine.md
Comment (dankenigsberg) [inline docs/quarantine.md:45]:
```suggestion
`[owner:@sig-{compute,network,storage}]` to each test's description. Also, a
```

File: docs/quarantine.md
Comment (dankenigsberg) [inline docs/quarantine.md:46]: We cannot include a downstream resources like this in an upstream doc. You can still use it if you please, but it cannot be part of a formal KubeVirt procedure. How about using https://github.com/features/project-management/ ?

File: docs/quarantine.md
Comment (davidvossel) [inline docs/quarantine.md:42]: Can we document a procedure for [release-blocker] tests which show up in this quarantine list? We can consider quarantining release-blocker tests, but they have a urgent priority and carry much larger risk. 

My suggestion would be something like this.

* create github issue with a comment ```/release-blocker master``` to ensure that the issue is addressed before a new release is cut.
* Ensure that the github issue is assigned to an individual who will own bringing the blocker to completion within a quick time frame.

File: docs/quarantine.md
Comment (dankenigsberg) [reply docs/quarantine.md:40]: I think that this pr should be accompanied by an email, calling out to the owners of the suspected tests.

File: docs/quarantine.md
Comment (davidvossel) [inline docs/quarantine.md:46]: 1. is there a reason not to use github for this? kanban boards exist that use github issues. https://docs.github.com/en/github/managing-your-work-on-github/about-project-boards 
2. is this board public for both viewing and contributing comments to?

**Review by dankenigsberg — CHANGES_REQUESTED** [review]
Thanks for this proposal, here's an initial review.

**Review by davidvossel — COMMENTED** [review]
looks reasonable. I had some comments about where to host the issues and how to handle release-blocker tests

File: docs/quarantine.md
Comment (davidvossel) [reply docs/quarantine.md:46]: also, for the ticket creation, not all tests will have test_id's associated with them. just pointing out that this isn't a reliable identifier

File: docs/quarantine.md
Comment (fgimenez) [reply docs/quarantine.md]: I added the second point as a consequence of the first and more focused on the status of the product and not the status of the suite. I've rephrased it a bit to make it more clear, let me know if we should merge them.

File: docs/quarantine.md
Comment (fgimenez) [reply docs/quarantine.md]: Sure, done.

File: docs/quarantine.md
Comment (fgimenez) [reply docs/quarantine.md:41]: Good point, added a note about it

File: docs/quarantine.md
Comment (fgimenez) [reply docs/quarantine.md]: Makes total sense to me, I've created this project https://github.com/kubevirt/kubevirt/projects/2?add_cards_query=is%3Aopen and linked it from the doc, let me know wdyt.

File: docs/quarantine.md
Comment (fgimenez) [reply docs/quarantine.md]: Indeed, added it to the doc, PTAL.

File: docs/quarantine.md
Comment (fgimenez) [reply docs/quarantine.md]: As commented above I would prefer to use https://github.com/kubevirt/kubevirt/projects/2?add_cards_query=is%3Aopen will take a look at the link you shared for managing gh issues from jira kanban boards, thx!

**Review by dhiller — APPROVED** [review]

**Comment by dhiller** [comment]
FYI: Further changes in the repo configs need to be done by creating a PR against orgs.yaml. Otherwise the periodic job will restore the old configuration again. See this PR as example: kubevirt/project-infra#1035

File: docs/quarantine.md
Comment (fgimenez) [reply docs/quarantine.md]: In order to prevent duplication of efforts I've removed references to tickets and boards, PTAL.

Comment (fgimenez) [reply docs/quarantine.md]: I finally removed references to tickets and boards, haven't found an easy way to have gh and jira boards in sync, let me know WDYT.

File: docs/quarantine.md
Comment (davidvossel) [inline docs/quarantine.md:49]: missing link here

Comment (fgimenez) [reply docs/quarantine.md]: Yep this was a leftover reference to the board, removed.

File: docs/quarantine.md
Comment (dhiller) [inline docs/quarantine.md:48]: How about we create teams for these sigs here: https://github.com/kubevirt/project-infra/blob/master/github/ci/prow-deploy/files/orgs.yaml#L651
This would have the advantage that we could assign the team directly on an issue or pr?

File: docs/quarantine.md
Comment (dhiller) [inline docs/quarantine.md:48]: How about we create teams for these sigs here: https://github.com/kubevirt/project-infra/blob/master/github/ci/prow-deploy/files/orgs.yaml#L651
This would have the advantage that we could assign the team directly on an issue or pr?

Comment (fgimenez) [reply docs/quarantine.md:48]: Great idea, this would make things much easier for tracking on the gh side, we can assign the issues from flake-issue-creator to these teams.

File: docs/quarantine.md
Comment (davidvossel) [inline docs/quarantine.md:39]: @fgimenez I've been thinking about these thresholds a little more. Depending on the sample size of test results, they may or may not be accurate for us.

For example, 2 failures over two weeks for a sample size of 200 test runs is a 1% failure rate... that's probably not worth us quarantining

However, 2 failures over two weeks with a sample size of 30 test runs is a something like a 6% failure rate, which to me crosses a threshold of needing attention.

What if we consider a failure percentage as our threshold rather than number of times failed? For me, I'd consider a 95% pass rate (5% failure rate) as a reasonable threshold as long as we have a sample size over 100.

**Review by davidvossel — COMMENTED** [review]
i made some comments about using failure percentages instead of hard numbers to more accurately determine flakiness

File: docs/quarantine.md
Comment (rmohr) [reply docs/quarantine.md:39]: Something like this would really be preferrable for me.

A few more considerations:
 * I would also suggest to not use jobs from PRs which were not merged. This means that we can only take what flakefinder already takes (presumbit runs on then finally merged code) + periodic jobs. I think we should look into creating a reasonable mass of runs with these.
  * I would also consider a grace period for tests with a certain percentage (of course not if they fail e.g. 20% of the time). There should be the chance to fix the tests within e.g. one or two days before they are moved to quarntine, once one notices that tests need attention. Otherwise one would have to wait at least a week to unquarantine them via automated processes, which leave blind spots.

Comment (fgimenez) [reply docs/quarantine.md:39]: > For me, I'd consider a 95% pass rate (5% failure rate) as a reasonable threshold as long as we have a sample size over 100.

totally agree, will update the doc accordingly.

> * I would also suggest to not use jobs from PRs which were not merged. This means that we can only take what flakefinder already takes (presumbit runs on then finally merged code) + periodic jobs.

makes sense to me, this removes a lot of noise.

> * I would also consider a grace period for tests with a certain percentage

This is very interesting, similar to what we have already in place for release blockers. 2 days of grace period since the quarantine PR is proposed sounds very good to me, I'll add that to the doc.

Comment (fgimenez) [reply docs/quarantine.md:39]: > Otherwise one would have to wait at least a week to unquarantine them via automated processes, which leave blind spots.

In any case we need to keep in mind that the test is still giving feedback, we keep executing it in the quarantined lanes (for now only periodics).

Comment (rmohr) [reply docs/quarantine.md:39]: > In any case we need to keep in mind that the test is still giving feedback, we keep executing it in the quarantined lanes (for now only periodics).

For that one has to know that a relevant test for a code change exists and that it was moved to quarantine. Further we expect then that people start looking into lanes where things are almost "supposed" to fail, while at the same time saying we move flaky tests out as fast as possible to avoid that people find a flaky test not relevant.

I find it great that we run them, and experienced maintainers can maybe still make something out of them, but I would not bring it up here as an argument that we still get feedback.

Comment (fgimenez) [reply docs/quarantine.md]: Ok pushed changes to address your comments @davidvossel @rmohr PTAL.

For the grace period I've included a number of successful executions for taking a fix into account, let me know wdyt.

Comment (davidvossel) [reply docs/quarantine.md]: > Ok pushed changes to address your comments @davidvossel @rmohr PTAL.

yep, new wording seems reasonable. I'd consider all these values as a starting point for us. Lets remain remain flexible to adjust these if we find our percentages or time frames don't end up serving us well.

Comment (dankenigsberg) [reply docs/quarantine.md]: >>  For me, I'd consider a 95% pass rate (5% failure rate) as a reasonable threshold as long as we have a sample size over 100.
> totally agree, will update the doc accordingly.

@davidvossel , @fgimenez : to me, a 95% pass rate can be too low for a single test. With 16 tests currently quarantined, it puts us at 95%**16=44% pass rate for the entire suite. e2e suite failure must be rare, so that when it happens everyone is startled and goes to check why did it happen. After some 5 weeks below 50%, we are on 80% today according to [grafana](https://grafana-kubevirt-prow.apps.ovirt.org/d/Cz6zgEEGz/e2e-jobs-overview?orgId=1&refresh=2h&from=now-14d&to=now&var-job_name=periodic-kubevirt-e2e-k8s-.*(prev%7Clatest)). I'd consider a 95% pass rate as the bare minimum for the entire per-pull suites, which means a pass rate higher than 99.5% for an individual test.

Comment (rmohr) [reply docs/quarantine.md]: Effectively, when looking at the whole quarantine flow:
 * test get into quarantine
 * someone gets assigned to fix the test
 * The person needs X days to find the issue
 * The test needs to succeed for one week from there on
 
how do we prevent having coverage holes for weeks on PRs?

Comment (dankenigsberg) [reply docs/quarantine.md]: To me, the interesting parameter is not an individual test pass rate but the total pass rate of the suites we run per PR. When tests fail on a PR, the PR author must be 95% sure that it their fault. Otherwise they blindly /retest and waste CI resources. 

> how do we prevent having coverage holes for weeks on PRs?

Suggestion: during the X+7 days of quarantine, the person assigned to the test must keep testing it on master. They must raise the alarm if suddenly their test fails consistently, and have the faulty PR reverted.

Comment (fgimenez) [reply docs/quarantine.md]: > Effectively 99.5% means a lot of tests are candidates for quarantine.

In fact only the tests that are actually preventing PRs from being merged because of non-determinism, if a test behaves properly then it won't be proposed even with this lower failure percentage.

> how do we prevent having coverage holes for weeks on PRs?

ATM we have the periodics that run the whole suite (including quarantined tests) 3 times per day to catch any error that went unnoticed in PRs, as soon as resources are in place we can run quarantine lanes also on PRs. In any case I wonder what coverage gives us a flaky test, we can't trust the information it provides. In my opinion proper test coverage comes from deterministic tests.

Comment (fgimenez) [reply docs/quarantine.md]: > > how do we prevent having coverage holes for weeks on PRs?
> 
> Suggestion: during the X+7 days of quarantine, the person assigned to the test must keep testing it on master. They must raise the alarm if suddenly their test fails consistently, and have the faulty PR reverted.

I think we should have this as part of the stability check for getting tests out of quarantine. After the fix has landed, the assigned person needs to check daily if the quarantine test executions from master (in periodics) have any failure. When periodics are integrated into flakefinder this will be easier to check.

Comment (fgimenez) [reply docs/quarantine.md]: > I'd consider a 95% pass rate as the bare minimum for the entire per-pull suites, which means a pass rate higher than 99.5% for an individual test.

Agree and updated accordingly, PTAL.

File: docs/quarantine.md
Comment (dankenigsberg) [inline docs/quarantine.md:41]: This may be too low to measure: it means one failure for 200 executions. When would one obtain such statistics?

I would not mind to keep a test that failed twice over two weeks if nothing else is failing. Can we ask to quarantine the most frequent fails, until the pass rate goes above 95% ?

Comment (davidvossel) [reply docs/quarantine.md:41]: > I would not mind to keep a test that failed twice over two weeks if nothing else is failing. Can we ask to quarantine the most frequent fails, until the pass rate goes above 95% ?

@dankenigsberg I think this is absolutely a great goal, but in my opinion optimizing for a 95% total test suite pass rate (rather than 95% test case pass rate) right now is too aggressive. To hit a 95% total test suite pass rate today, we might have to disable so many tests that we begin to expose the project to further instability. It's the whole "we're working on an airplane in flight" analogy, where if we are too aggressive we risk falling out of the sky. 

I'd like to see 95% total test suite pass rate as our goal, and our quarantine policy to stick with the previously mentioned "quarantine a test case after x% of failures over y time". We will start to see improvements. For example, this threshold would definitely have caught the issues that caused the merge issues two weeks ago. There we had test cases failing more like 10-30% of the time.

Over time, we should begin tweaking this per test case percentage with the goal of getting it down to 1% and even 0.5% per test case failure rates. I believe that will eventually hit your 95% total test suite pass rate, but in an incremental way that introduces less risk.

thoughts?

File: docs/quarantine.md
Comment (fgimenez) [reply docs/quarantine.md]: > For that one has to know that a relevant test for a code change exists and that it was moved to quarantine. Further we expect then that people start looking into lanes where things are almost "supposed" to fail, while at the same time saying we move flaky tests out as fast as possible to avoid that people find a flaky test not relevant.
> 
> I find it great that we run them, and experienced maintainers can maybe still make something out of them, but I would not bring it up here as an argument that we still get feedback.

As I see it this is where test owners come into play, they need to take care of bringing the test out of quarantine and should be aware of any failures during the quarantine period. 

I agree that keeping track of all the executions on all the PRs (to a lesser extent also all the periodic executions) can be hard, but we could help here with automation, maybe drop a comment on a PR that had a failing quarantine test tagging the owner, or sending an email to the owner after a quarantined test failed on a periodic, wdyt?

File: docs/quarantine.md
Comment (fgimenez) [reply docs/quarantine.md]: From my POV it would be better to set rates as goals instead of concrete numbers without references to how many executions were done. I also think that this is an initial policy that should not be set on stone, we should adapt it to our needs over time, we can modify the concrete numbers as needed.

I agree that 0.5% is too low to measure (with the current execution rate, just one failure would represent a much higher failure rate), I've reverted it to 5%, let me know if that is ok.

Comment (dankenigsberg) [reply docs/quarantine.md]: @davidvossel I'm all for gradual improvements, not pies-in-the-skies. I'd like us all to never accept again a condition where doc PRs fail more than 5% of the times.

for example, this very PR has
```
Test name | Commit | Details | Rerun command
-- | -- | -- | --
pull-kubevirt-e2e-k8s-1.19 | e57c442 | link | /test pull-kubevirt-e2e-k8s-1.19
pull-kubevirt-e2e-k8s-1.18 | e57c442 | link | /test pull-kubevirt-e2e-k8s-1.18
pull-kubevirt-unit-test | e5a5c4f | link | /test pull-kubevirt-unit-test
```
It makes no sense at all. Our procedures must ensure that people are working to fix this. To me, this quarantine document is a mechanism to pull in many people to work on this critical point in our project. How can we ensure that people focus on CI improvement until the failure rate falls below 5%?

Comment (fgimenez) [reply docs/quarantine.md]: I've added an additional section to the doc with the goal of the methodology set to whole test suite failure rate at 10% (for considering the suite as stable) and per test failure rate to 5% (for the concrete test to be proposed for quarantine), let me know wdyt.

Comment (dankenigsberg) [reply docs/quarantine.md]: @davidvossel, I understand. But I would like this document to express that having 17 individual tests each failing 4% of the time we get to 96%**17=50% fail rate, **requires** an aggressive stance.

Comment (fgimenez) [reply docs/quarantine.md]: @dankenigsberg I've added a note about how the whole suite passing rate depends on the individual tests passing rate, PTAL.

File: docs/quarantine.md
Comment (EdDev) [inline docs/quarantine.md:78]: What is this "tracking ticket"? I do not think it was mentioned earlier, except for the blockers case which was talking about a github issue. But here it seems generic, so perhaps there is a need to specify when such an issue is to be opened.

Another point is about the owner: The text and in other places talked about an individual, but I would prefer to see here a group/SIG. As we mark ownership per SIG on the tests, I think we should do the same when talking about ownership. I guess we can mix this (SIG or person) if you think that level of granularity is needed.

File: docs/quarantine.md
Comment (EdDev) [inline docs/quarantine.md:90]: A bit too long sentence, can you split it?

**Review by EdDev — CHANGES_REQUESTED** [review]
Thank you, it looks very nice.

There is one thing that I found missing: What happens if an quarantine is not handled and rots? There needs to be an escalation and actions taken in order to avoid cases where tests are left in the quarantine for too long.

File: docs/quarantine.md
Comment (fgimenez) [reply docs/quarantine.md]: Absolutely, I've rewritten all of the references to how we calculate failures, let me know wdyt now.

File: docs/quarantine.md
Comment (fgimenez) [reply docs/quarantine.md]: Makes sense to me, updated.

File: docs/quarantine.md
Comment (fgimenez) [reply docs/quarantine.md:58]: I think this is out of the scope of this doc, we can change the implementation later. At the end we want to somehow mark the test in its description, and this is what this doc should care about at this stage IMO, what do you think?

File: docs/quarantine.md
Comment (fgimenez) [reply docs/quarantine.md:48]: Sure, done.

File: docs/quarantine.md
Comment (fgimenez) [reply docs/quarantine.md]: Yes, this "tracking ticket" is a leftover, removed. I agree also that the owner should be always the team, the "person" references also came from when we described management of tickets here, I've replaced all references person -> team.

File: docs/quarantine.md
Comment (fgimenez) [reply docs/quarantine.md]: Sure done.

**Review by EdDev — COMMENTED** [review]
Thank you, it look good to me.
I'll wait for the review commits to be squashed to LGTM this.

**Comment by fgimenez** [comment]
> I'll wait for the review commits to be squashed to LGTM this.

Thanks! Ok, commits are already squashed.

**Comment by fgimenez** [comment]
/retest

```
docker: Error response from daemon: Get "https://quay.io/v2/": net/http: request canceled (Client.Timeout exceeded while awaiting headers).
```
but no incident reported in https://status.quay.io/#past-incidents maybe a connectivity issue.

**Review by EdDev — APPROVED** [review]
Thanks!

**Comment by fgimenez** [comment]
wdyt @davidvossel is there anything missing here?

**Review by davidvossel — APPROVED** [review]
/approve

**Pull request merged**
